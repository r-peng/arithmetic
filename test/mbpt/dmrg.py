import numpy as np
from pyblock3.hamiltonian import Hamiltonian
from pyblock3.fcidump import FCIDUMP
from pyblock3.algebra.mpe import MPE
from pyblock3.symbolic.expr import OpElement, OpNames
from pyblock3.algebra.symmetry import SZ
norb = 6
na = norb // 2
nb = na

t_leads = 1.
t_hyb = .4
U = 1.
idot = norb // 2

h1 = np.zeros((norb,norb))
for i in range(norb):
    if i < norb - 1:
        dot = (i == idot or i + 1 == idot)
        h1[i, i + 1] = -t_hyb if dot else -t_leads
    if i > 0:
        dot = (i == idot or i - 1 == idot)
        h1[i, i - 1] = -t_hyb if dot else -t_leads
w,v = np.linalg.eigh(h1)
print('E0=',sum(w[:na])*2)
rdm = np.einsum('pi,qi->pq',v[:,:na],v[:,:na])
for i in range(norb):
    print(i,rdm[i,i]*2)
def build_qc(u=2, t_leads=1, t_hyb=1, idot=None, n=8, cutoff=1E-9):
    fcidump = FCIDUMP(pg='c1', n_sites=n, n_elec=n, twos=0, ipg=0, orb_sym=[0] * n)
    hamil = Hamiltonian(fcidump, flat=False)
    idot = n // 2 if idot is None else idot
    def generate_terms(n_sites, c, d):
        for i in range(0, n_sites):
            if i<norb-1:
                dot = (i == idot or i + 1 == idot)
                fac = -t_hyb if dot else -t_leads
                for s in [0, 1]:
                    yield fac * c[i, s] * d[i + 1, s]
            if i > 0:
                dot = (i == idot or i - 1 == idot)
                fac = -t_hyb if dot else -t_leads
                for s in [0, 1]:
                    yield fac * c[i, s] * d[i - 1, s]
        if abs(u)>0:
            yield u * (c[idot, 0] * c[idot, 1] * d[idot, 1] * d[idot, 0])
    return hamil, hamil.build_mpo(generate_terms, cutoff=cutoff).to_sparse()

hamil, mpo = build_qc(u=0,t_leads=t_leads,t_hyb=t_hyb, idot=idot, n=norb)
mpo, _ = mpo.compress(left=True, cutoff=1E-9, norm_cutoff=1E-9)
mpo = mpo.to_flat()
bdim = 500
mps = hamil.build_mps(bdim)
mps = mps.to_flat()
noises = [1E-4, 1E-5, 1E-6, 0]
davthrds = None
dmrg = MPE(mps, mpo, mps).dmrg(bdims=[bdim], noises=noises,
                            dav_thrds=davthrds, iprint=0, n_sweeps=20, tol=1E-12)
for i in range(norb):
    dop = OpElement(OpNames.D, (i, 0), q_label=SZ(-1, -1, hamil.orb_sym[i]))
    di = hamil.build_site_mpo(dop)
    di = di.to_flat()
    n_expt = np.dot(mps.conj(), mps)
    q = 2 * np.dot((di @ mps).conj(), di @ mps)
    print(f'i={i},norm={n_expt},q={q/n_expt}')
hamil, mpo = build_qc(u=U,t_leads=t_leads,t_hyb=t_hyb,idot=idot,n=norb)
mpo, _ = mpo.compress(left=True, cutoff=1E-9, norm_cutoff=1E-9)
mpo = mpo.to_flat()

isite = idot
dop = OpElement(OpNames.D, (isite, 0), q_label=SZ(-1, -1, hamil.orb_sym[isite]))
di = hamil.build_site_mpo(dop)
di = di.to_flat()

mpo.const -= dmrg.energies[-1]
mpe = MPE(mps,mpo,mps)
dt = 0.1
n_steps = 10
for it in range(n_steps):
    cur_t = (it + 1) * dt
    te = mpe.tddmrg(bdims=[bdim], dt=-dt*1j, iprint=0, n_sweeps=1, n_sub_sweeps=2, normalize=False)
    n_expt = np.dot(mps.conj(), mps)
    q = 2 * np.dot((di @ mps).conj(), di @ mps)
    print(f't={cur_t},norm={n_expt},q={q/n_expt}')

