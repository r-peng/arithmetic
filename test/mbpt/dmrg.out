E0= -5.560960262062946
0 0.9999999999999996
1 1.000000000000001
2 0.9999999999999999
3 1.0000000000000004
4 1.0
5 1.000000000000001
MPO site    0 /    6
MPO site    1 /    6
MPO site    2 /    6
MPO site    3 /    6
MPO site    4 /    6
Time elapsed =      0.126 | E =   -5.560960260485841 | DE = 0.00E+00 | MDW = 8.13E-15 | MEM = 60.2 MB
Time sweep =      0.126 | Time davidson =      0.000 | Time decomp =      0.025
Time elapsed =      0.177 | E =   -5.560960262044967 | DE = 1.56E-09 | MDW = 1.09E-14 | MEM = 60.4 MB
Time sweep =      0.050 | Time davidson =      0.000 | Time decomp =      0.023
Time elapsed =      0.226 | E =   -5.560960262062833 | DE = 1.79E-11 | MDW = 4.13E-15 | MEM = 60.5 MB
Time sweep =      0.050 | Time davidson =      0.000 | Time decomp =      0.019
Time elapsed =      0.282 | E =   -5.560960262062948 | DE = 1.15E-13 | MDW = 1.44E-15 | MEM = 60.5 MB
Time sweep =      0.056 | Time davidson =      0.000 | Time decomp =      0.003
i=0,norm=1.0000000000000007,q=1.0000000000093106
i=1,norm=1.0000000000000007,q=0.9999999999991526
i=2,norm=1.0000000000000007,q=1.0000000000067764
i=3,norm=1.0000000000000007,q=0.9999999999872456
i=4,norm=1.0000000000000007,q=1.0000000000002274
i=5,norm=1.0000000000000007,q=0.9999999999972871
MPO site    0 /    6
MPO site    1 /    6
MPO site    2 /    6
MPO site    3 /    6
MPO site    4 /    6
t=0.1,norm=(0.9999999999990694+1.1057766451194274e-18j),q=(0.997861787222926+7.277239723322575e-20j)
t=0.2,norm=(0.9999999999981156-3.804593281205152e-18j),q=(0.9915149355017725-4.05893205202268e-18j)
t=0.30000000000000004,norm=(0.9999999999971367+3.812660977283361e-18j),q=(0.9811590813216666+2.420490716472401e-18j)
t=0.4,norm=(0.9999999999961839-5.385327066283809e-18j),q=(0.9671148526592795+7.664625339297815e-18j)
t=0.5,norm=(0.9999999999952038-1.0788050981620527e-17j),q=(0.949806723992232-2.3035055202512274e-17j)
t=0.6000000000000001,norm=(0.9999999999942228-3.7428882538404217e-17j),q=(0.9297409489175898-1.613716151504963e-17j)
t=0.7000000000000001,norm=(0.9999999999932603+4.231048737242226e-18j),q=(0.9074802404192076-9.280932778522137e-18j)
t=0.8,norm=(0.999999999992307+3.825240494469823e-18j),q=(0.8836170498611625+2.880720427515215e-17j)
t=0.9,norm=(0.9999999999913529+2.2560463228996254e-17j),q=(0.858747265689635+1.0353732206334366e-17j)
t=1.0,norm=(0.9999999999904012-3.873723204532781e-18j),q=(0.8334459891401276-1.246528737899044e-17j)
